### DISCUSSION TITLE
How do decode attachment file name with Quoted-Printable .

### DISCUSSION THREAD

**Discussion by shenglingshi** [jstedfast/MimeKit#985] [Q&A]
when a email attachment filename is using  Quoted-Printable encode.
when i save attachment ,it's not original filename.

**Comment by jstedfast** [comment]
I would need to see your code and an example that illustrates the behavior that you are seeing because it works perfectly fine for me.

Also, do you mean the *name* of the file isn't correct? Or do you mean the *content* isn't correct? Your question is so vague I'm not even sure what you mean.

**Reply by shenglingshi** [reply]
Sorry for my vague question. And Thank for you answe my question.
when I save attachment with this code:

```
foreach (var attachment in message.Attachments) {
    if (attachment is MessagePart) {
        var fileName = attachment.ContentDisposition?.FileName;
        var rfc822 = (MessagePart) attachment;

        if (string.IsNullOrEmpty (fileName))
            fileName = "attached-message.eml";

        using (var stream = File.Create (fileName))
            rfc822.Message.WriteTo (stream);
    } else {
        var part = (MimePart) attachment;
        var fileName = part.FileName;

        using (var stream = File.Create (fileName))
            part.Content.DecodeTo (stream);
    }
}
```

the fileName is  Quoted-Printable encoded
and encoded filename like this "546w5Zyo57uZ5a6i5oi35YWN6LS5?= =?UTF-8?B?6K+V5Y2V5LiN5pS25Y+W5Lu75L2V6LS555SoLnBuZw==?=";
and I also find Api have a QuotedPrintableDecoder Class.
But I don't Know how to use it.

**Reply by jstedfast** [reply]
MimeKit decodes the fileName for you. The string you pasted is also not QuotedPrintable encoded, it is Rfc2047 encoded.

That said, it also appears corrupt which is probably why MimeKit did not decode it.

Can you look at the raw MiME headers and paste the Content-Disposition value for that MIME part?

I want to check if MimeKit is doing anything wrong or if it really is corrupt.

**Reply by shenglingshi** [reply]
Ok ,I export email file raw data to txt. 
You can debug it.
[email.raw.data.txt](https://github.com/jstedfast/MimeKit/files/13960415/email.raw.data.txt)

**Reply by jstedfast** [reply]
It looks like the headers really are corrupt:

```
Content-Type: image/png; 
	name*0="546w5Zyo57uZ5a6i5oi35YWN6LS5?=
 =?UTF-8?B?6K+V5Y2V5LiN5pS25Y+"; 
	name*1="W5Lu75L2V6LS555SoLnBuZw==?="
Content-Transfer-Encoding: base64
Content-Disposition: attachment; 
	filename*0="546w5Zyo57uZ5a6i5oi35YWN6LS5?=
 =?UTF-8?B?6K+V5Y2V5LiN5pS25Y+"; 
	filename*1="W5Lu75L2V6LS555SoLnBuZw==?="
```

This is incorrectly encoded for multiple reasons:

1. The client that created this tried to mix & match rfc2231 encoding with rfc2047 encoding which is absolutely bizarre.
2. The first rfc2047 encoded-word token (of the name and filename values) is missing the leading `=?UTF-8?B?` (or, at least, I presume that's what it intended to do).

If you edit the message and modify those headers to look like this, it should work:

```
Content-Type: image/png; 
	name="=?UTF-8?B?546w5Zyo57uZ5a6i5oi35YWN6LS5?= =?UTF-8?B?6K+V5Y2V5LiN5pS25Y+W5Lu75L2V6LS555SoLnBuZw==?="
Content-Transfer-Encoding: base64
Content-Disposition: attachment; 
	filename="=?UTF-8?B?546w5Zyo57uZ5a6i5oi35YWN6LS5?= =?UTF-8?B?6K+V5Y2V5LiN5pS25Y+W5Lu75L2V6LS555SoLnBuZw==?="
```

This will get you a filename of: 现在给客户免费试单不收取任何费用.png

**Reply by shenglingshi** [reply]
Ok , Thanks.
